crates/primitives/src/transaction/signature.rs

```diff
@@ -17,41 +17,34 @@ pub struct Signature {
 }
 
 impl Signature {
-    /// Encode the `v`, `r`, `s` values without a RLP header.
-    /// Encodes the `v` value using the legacy scheme without EIP-155.
-    pub(crate) fn encode_inner_legacy(&self, out: &mut dyn reth_rlp::BufMut) {
-        (self.odd_y_parity as u8 + 27).encode(out);
-        self.r.encode(out);
-        self.s.encode(out);
-    }
-
     /// Output the length of the signature without the length of the RLP header, using the legacy
-    /// scheme without EIP-155.
-    pub(crate) fn payload_len_legacy(&self) -> usize {
-        (self.odd_y_parity as u8 + 27).length() + self.r.length() + self.s.length()
+    /// scheme with EIP-155 support depends on chain_id.
+    pub(crate) fn payload_len_legacy(&self, chain_id: Option<u64>) -> usize {
+        self.v(chain_id).length() + self.r.length() + self.s.length()
     }
 
     /// Encode the `v`, `r`, `s` values without a RLP header.
-    /// Encodes the `v` value with EIP-155 support, using the specified chain ID.
-    pub(crate) fn encode_eip155_inner(&self, out: &mut dyn reth_rlp::BufMut, chain_id: u64) {
-        // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
-        let v = chain_id * 2 + 35 + self.odd_y_parity as u64;
-        v.encode(out);
+    /// Encodes the `v` value using the legacy scheme with EIP-155 support depends on chain_id.
+    pub(crate) fn encode_legacy(&self, out: &mut dyn reth_rlp::BufMut, chain_id: Option<u64>) {
+        self.v(chain_id).encode(out);
         self.r.encode(out);
         self.s.encode(out);
     }
 
-    /// Output the length of the signature without the length of the RLP header, with EIP-155
-    /// support.
-    pub(crate) fn eip155_payload_len(&self, chain_id: u64) -> usize {
-        // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
-        let v = chain_id * 2 + 35 + self.odd_y_parity as u64;
-        v.length() + self.r.length() + self.s.length()
+    /// Output the `v` of the signature depends on chain_id
+    #[inline]
+    fn v(&self, chain_id: Option<u64>) -> u64 {
+        if let Some(chain_id) = chain_id {
+            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
+            self.odd_y_parity as u64 + chain_id * 2 + 35
+        } else {
+            self.odd_y_parity as u64 + 27
+        }
     }
 
     /// Decodes the `v`, `r`, `s` values without a RLP header.
     /// This will return a chain ID if the `v` value is EIP-155 compatible.
-    pub(crate) fn decode_eip155_inner(buf: &mut &[u8]) -> Result<(Self, Option<u64>), DecodeError> {
+    pub(crate) fn decode_legacy(buf: &mut &[u8]) -> Result<(Self, Option<u64>), DecodeError> {
```

Comment:
same nit re: legacy in the method name

Reply:
I think we need to keep `_legacy` name for `Legacy` transaction.

Reply:
I'm also not a fan of this rename here, because this makes it less obvious that this supports EIP-155.
perhaps we can combine the two, eip155 and legacy?

Reply:
@mattsse Thanks for your feedback! The below is my thoughts.

* `transaction/signature.rs` is a part of `transaction`.
* `Signature.encode() / decode()` works w/ owner `Transaction` and depends on it's type.
* `Transaction` types are Legacy,  EIP2930 and EIP1559. As a broader category, Legacy and Non Legacy (EIP2930, EIP1559).
* Legacy transaction could be EIP155 or not.

So I'd like to keep the current name. But I'm still thinking more for your opinion.

Reply:
@jinsankim thanks for the thoughts.. I liked @mattsse's suggestion of naming them `encode_with_eip155_chain_id` and `decode_with_eip155_chain_id`. do you mind changing them to that name? if that sounds good, we'll merge the PR after that

Reply:
@gakonst Thanks for your opinion and also, of course, I open to rename but I'd like to hear more opinion for `payload_len_legacy()`. You might intend to rename it also as `payload_len_with_eip155_chain_id()`, right?

In terms of encoding interface, `Signature` has 3 kinds of methods, `payload_len()`, `encode()` and `decode()`. Considering `Legacy` and `Non-Legacy`. `Signature` has currently 6 methods.

```
payload_len_legacy()
encode_legacy()
decode_legacy()

payload_len()
encode()
decode()
```

Even though I open to rename it but actually I'm little bit afraid to lost `legacy` term because these method used for `Legacy` transactions which follows EIP155 or not.

Reply:
https://github.com/paradigmxyz/reth/blob/fab15089546022a931f2e24763c902714ec137f4/crates/primitives/src/transaction/mod.rs#L166-L173

Reply:
https://github.com/paradigmxyz/reth/blob/fab15089546022a931f2e24763c902714ec137f4/crates/primitives/src/transaction/mod.rs#L18-L22

Reply:
https://github.com/paradigmxyz/reth/blob/fab15089546022a931f2e24763c902714ec137f4/crates/primitives/src/transaction/mod.rs#L60-L64

Reply:
https://github.com/paradigmxyz/reth/blob/fab15089546022a931f2e24763c902714ec137f4/crates/primitives/src/transaction/mod.rs#L108-L112

Reply:
> You might intend to rename it also as payload_len_with_eip155_chain_id()

this is backward because `Signature::payload_len_legacy` is only used if there's no chain id.

`payload_len_legacy` is probably fine here, we could rename this to `pre_eip155_payload_len` or something.

Reply:
All type of transactions have chain_id follow EIP155 up in order to protect replay attack.

Reply:
this is not correct for legacy tx where chain id is optional

> pub chain_id: Option<ChainId>, 

tbh, I'm a bit confused about what we're actually arguing here.

Reply:
https://github.com/paradigmxyz/reth/pull/928#issuecomment-1398617078

Reply:
> tbh, I'm a bit confused about what we're actually arguing here.

I've got feedback to rename `...with_eip155_chain_id` from gakonst as well. I open to rename it but I'm little bit afraid to lost `legacy` from function names. So I asked it whether it's really ok without `legacy`.

Reply:
Just another idea, using `_with_legacy_tx()` as postfix.

Reply:
`legacy` has no real meaning for `Signature`, because a `legacy` tx can still support eip155.
so the functions should instead use the term `eip155` or `chain_id` to distinguish how the sig of a legacy tx is encoded. These functions are called conditionally based on whether the legacy has a chain id or not.